pythonwhat/probe.py

```diff
@@ -0,0 +1,218 @@
+
+    def __iter__(self):
+        for c in self.child_list: yield c
+
+class NodeList(Node):
+    def partial(self):
+        return [node.partial() for node in self.child_list]
+
+    def update_child_calls(self):
+        pass
+
+class NodeDict(Node):
+    def partial(self):
+        return OrderedDict((node.arg_name, node.partial()) for node in self.child_list)
+
+    def update_child_calls(self):
+        pass
+
+class Probe(object):
+    def __init__(self, tree, f):
+        self.tree = tree
+        self.f = f
+        self.test_name = f.__name__
+        # TODO: auto sub_test detection
+        self.sub_tests = SUB_TESTS.get(self.test_name) or []
+    
+    def __call__(self, *args, **kwargs):
+        """Bind arguments to original function signature, and store in a Node
+
+        This is used to discover what tests and sub-tests the SCT would like to 
+        call, and defer them for later execution via their node instance. Node
+        instances are assembled into a tree.
+
+        """
+
+        bound_args = inspect.signature(self.f).bind(*args, **kwargs)
+
+        data = dict(
+                bound_args = bound_args, 
+                func = self.f)
+        this_node = Node(data=data, name=self.test_name)
+        self.tree.crnt_node.add_child(this_node)
+
+        da = bound_args.arguments
+        for st in self.sub_tests:     # TODO: auto sub test detection
+            if st in da and da[st]:
+                self.run_sub_tests(da[st], self.tree, this_node, st)
+        return this_node
+
+    @staticmethod
+    def run_sub_tests(test, tree, node, arg_name):
+        # note that I've made the strong assumption that
+        # if not a function, then test is a dict, list or tuple of them
+        if isinstance(test, dict):
+            nd = NodeDict(name = "Dict", arg_name = arg_name)
+            node.add_child(nd)
+            for k, f in test.items(): Probe.run_sub_tests(f, tree, nd, k)
+        elif isinstance(test, (list, tuple)): 
+            nl = NodeList(name = "List", arg_name = arg_name)
+            node.add_child(nl)
+            for ii, f in enumerate(test): Probe.run_sub_tests(f, tree, nl, str(ii))
+        elif isinstance(test, Node): 
+            # test was a lambdaless subtest call, which produced a node
+            # so need to tell it what its arg_name was on parent test
+            test.arg_name = arg_name
+            node.add_child(test)
+        elif callable(test): 
+            # test was inside a lambda or function containing subtests
+            # since either may contain multiple subtests, we put them in a node list
+            nl = NodeList(name = "ListDeferred", arg_name = arg_name)
+            node.add_child(nl)
+            prev_node, tree.crnt_node = tree.crnt_node, nl
+            test()
+            tree.crnt_node = prev_node
+        elif test is not None:
+            raise Exception("Expected a function or list/tuple/dict of functions")
+
+
+def create_test_probes(test_exercise):
```

Comment:
I find `test_exercise` a confusing argument here. Can you use a name that refers to the fact it is a context?